# **Waze Churn Analysis Project - Data Exploration**
This notebook focuses on the initial exploration of the raw user-level dataset used in the Waze churn analysis project. The objective is to understand the structure, quality, and key characteristics of the data prior to formal exploratory data analysis and modelling.

Specifically, this notebook will:
- Examine the dataset schema, data types, and basic distributions  
- Identify missing values, anomalies, and potential data quality issues  
- Highlight early behavioural patterns and observations relevant to user churn  
- Produce a cleaned and standardised dataset suitable for downstream EDA and modelling  

Key findings and observations are documented in **“Insight:”** sections following relevant analysis steps.

For detailed definitions of each variable, please refer to the data dictionary located at `docs/data_dictionary.md`.

## **1. Imports and Data Loading**
















In [1]:
# Import packages for data manipulation
import pandas as pd
import numpy as np

In [3]:
# Load raw dataset into dataframe
df = pd.read_csv('waze_dataset.csv') # Adjust the path according to your runtime structure

## **2. Summary Information**

In [4]:
# Show first 10 rows of the dataset to get general understanding of the data
df.head(10)

ID     label  sessions  drives  total_sessions  n_days_after_onboarding  \
0   0  retained       283     226      296.748273                     2276   
1   1  retained       133     107      326.896596                     1225   
2   2  retained       114      95      135.522926                     2651   
3   3  retained        49      40       67.589221                       15   
4   4  retained        84      68      168.247020                     1562   
5   5  retained       113     103      279.544437                     2637   
6   6  retained         3       2      236.725314                      360   
7   7  retained        39      35      176.072845                     2999   
8   8  retained        57      46      183.532018                      424   
9   9   churned        84      68      244.802115                     2997   

   total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
0                     208                       0       2628.845068   
1                      19                      64      13715.920550   
2                       0                       0       3059.148818   
3                     322                       7        913.591123   
4                     166                       5       3950.202008   
5                       0                       0        901.238699   
6                     185                      18       5249.172828   
7                       0                       0       7892.052468   
8                       0                      26       2651.709764   
9                      72                       0       6043.460295   

   duration_minutes_drives  activity_days  driving_days   device  
0              1985.775061             28            19  Android  
1              3160.472914             13            11   iPhone  
2              1610.735904             14             8  Android  
3               587.196542              7             3   iPhone  
4              1219.555924             27            18  Android  
5               439.101397             15            11   iPhone  
6               726.577205             28            23   iPhone  
7              2466.981741             22            20   iPhone  
8              1594.342984             25            20  Android  
9              2341.838528              7             3   iPhone

In [5]:
# Show basic information of the dataset and columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  object 
dtypes: float64(3), int64(8), object(2)
memory usage: 1.5+ MB


In [21]:
# Show descriptive statistics of numeric columns
df.describe()

ID      sessions        drives  total_sessions  \
count  14999.000000  14999.000000  14999.000000    14999.000000   
mean    7499.000000     80.633776     67.281152      189.964447   
std     4329.982679     80.699065     65.913872      136.405128   
min        0.000000      0.000000      0.000000        0.220211   
25%     3749.500000     23.000000     20.000000       90.661156   
50%     7499.000000     56.000000     48.000000      159.568115   
75%    11248.500000    112.000000     93.000000      254.192341   
max    14998.000000    743.000000    596.000000     1216.154633   

       n_days_after_onboarding  total_navigations_fav1  \
count             14999.000000            14999.000000   
mean               1749.837789              121.605974   
std                1008.513876              148.121544   
min                   4.000000                0.000000   
25%                 878.000000                9.000000   
50%                1741.000000               71.000000   
75%                2623.500000              178.000000   
max                3500.000000             1236.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count            14999.000000      14999.000000             14999.000000   
mean                29.672512       4039.340921              1860.976012   
std                 45.394651       2502.149334              1446.702288   
min                  0.000000         60.441250                18.282082   
25%                  0.000000       2212.600607               835.996260   
50%                  9.000000       3493.858085              1478.249859   
75%                 43.000000       5289.861262              2464.362632   
max                415.000000      21183.401890             15851.727160   

       activity_days  driving_days  
count   14999.000000  14999.000000  
mean       15.537102     12.179879  
std         9.004655      7.824036  
min         0.000000      0.000000  
25%         8.000000      5.000000  
50%        16.000000     12.000000  
75%        23.000000     19.000000  
max        31.000000     30.000000

**Insight:**

1. There are 14,999 rows and 13 columns. Each row represent one unique user, therefore this dataset contains information of 14,999 users.
2. The variables `label` and `device` are of type `object`; `total_sessions`, `driven_km_drives`, and `duration_minutes_drives` are of type `float64`; the rest of the variables are of type `int64`. It is interesting that the `total_sessions` is stored as decimals instead of round numbers.
3. The dataset has 700 missing values in the `label` column. The missing values account for ~4.7% of the total rows, indicating a relatively low number of
missing data. There are no missing value in other columns.
4. There are several extreme values on the upper end of the distributions. For example, the maximum value of the `sessions variable is 743, corresponding to an average of approximately 24 app sessions per day. Likewise, the maximum value in the `driven_km_drives` column is 21,183 km, which exceeds half of the Earth’s circumference. These values indicate a small number of users with usage patterns that are substantially higher than the majority of the dataset.

## **3. Distinct Values**
For the variables of type `object`, we'll check the available values and their proportion.

In [16]:
df['label'].value_counts(normalize=True)

label
retained    0.822645
churned     0.177355
Name: proportion, dtype: float64

In [17]:
df['device'].value_counts(normalize=True)

device
iPhone     0.644843
Android    0.355157
Name: proportion, dtype: float64

**Insight:**

1. The target variable `label` is binary, with values `retained` and `churned`. Approximately **18% of users in the dataset are labelled as churned**, indicating a moderate level of monthly churn within the observed population. The remaining 82% are retained.

2. The `device` variable consists of two categories: `iPhone` and `Android`. iPhone users account for roughly 64% of the dataset, while the remaining users are on Android, suggesting a platform skew toward iOS within this user sample.

##**4. Missing Values**
Next we'll investigate the missing values to see whether there is any systematic pattern of the missing data.

### Summary Statistics
Compare the summary statistics of the 700 rows that are missing labels with summary statistics of the rows that are not missing any values.

In [6]:
# Isolate rows with null values
null_df = df[df['label'].isnull()]
# Display summary stats of rows with null values
null_df.describe()

ID    sessions      drives  total_sessions  \
count    700.000000  700.000000  700.000000      700.000000   
mean    7405.584286   80.837143   67.798571      198.483348   
std     4306.900234   79.987440   65.271926      140.561715   
min       77.000000    0.000000    0.000000        5.582648   
25%     3744.500000   23.000000   20.000000       94.056340   
50%     7443.000000   56.000000   47.500000      177.255925   
75%    11007.000000  112.250000   94.000000      266.058022   
max    14993.000000  556.000000  445.000000     1076.879741   

       n_days_after_onboarding  total_navigations_fav1  \
count               700.000000              700.000000   
mean               1709.295714              118.717143   
std                1005.306562              156.308140   
min                  16.000000                0.000000   
25%                 869.000000                4.000000   
50%                1650.500000               62.500000   
75%                2508.750000              169.250000   
max                3498.000000             1096.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count              700.000000        700.000000               700.000000   
mean                30.371429       3935.967029              1795.123358   
std                 46.306984       2443.107121              1419.242246   
min                  0.000000        290.119811                66.588493   
25%                  0.000000       2119.344818               779.009271   
50%                 10.000000       3421.156721              1414.966279   
75%                 43.000000       5166.097373              2443.955404   
max                352.000000      15135.391280              9746.253023   

       activity_days  driving_days  
count     700.000000    700.000000  
mean       15.382857     12.125714  
std         8.772714      7.626373  
min         0.000000      0.000000  
25%         8.000000      6.000000  
50%        15.000000     12.000000  
75%        23.000000     18.000000  
max        31.000000     30.000000

In [7]:
# Isolate rows without null values
not_null_df = df[~df['label'].isnull()]
# Display summary stats of rows without null values
not_null_df.describe()

ID      sessions        drives  total_sessions  \
count  14299.000000  14299.000000  14299.000000    14299.000000   
mean    7503.573117     80.623820     67.255822      189.547409   
std     4331.207621     80.736502     65.947295      136.189764   
min        0.000000      0.000000      0.000000        0.220211   
25%     3749.500000     23.000000     20.000000       90.457733   
50%     7504.000000     56.000000     48.000000      158.718571   
75%    11257.500000    111.000000     93.000000      253.540450   
max    14998.000000    743.000000    596.000000     1216.154633   

       n_days_after_onboarding  total_navigations_fav1  \
count             14299.000000            14299.000000   
mean               1751.822505              121.747395   
std                1008.663834              147.713428   
min                   4.000000                0.000000   
25%                 878.500000               10.000000   
50%                1749.000000               71.000000   
75%                2627.500000              178.000000   
max                3500.000000             1236.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count            14299.000000      14299.000000             14299.000000   
mean                29.638296       4044.401535              1864.199794   
std                 45.350890       2504.977970              1448.005047   
min                  0.000000         60.441250                18.282082   
25%                  0.000000       2217.319909               840.181344   
50%                  9.000000       3496.545617              1479.394387   
75%                 43.000000       5299.972162              2466.928876   
max                415.000000      21183.401890             15851.727160   

       activity_days  driving_days  
count   14299.000000  14299.000000  
mean       15.544653     12.182530  
std         9.016088      7.833835  
min         0.000000      0.000000  
25%         8.000000      5.000000  
50%        16.000000     12.000000  
75%        23.000000     19.000000  
max        31.000000     30.000000

**Insight:**
* Comparing summary statistics of the observations with missing retention labels with those that aren't missing
any values reveals nothing remarkable. The means and standard deviations are fairly consistent between the two groups.

### Null Values & Device Type

Next, check if there is a relation between null values and user's device.
Of the rows with null values, we'll calculate the percentage with each device (Android and iPhone) and then compared it with the percentage in the full dataset.


In [18]:
# Calculate % of iPhone nulls and Android nulls
null_df['device'].value_counts(normalize=True)

device
iPhone     0.638571
Android    0.361429
Name: proportion, dtype: float64

**Insight:**
* The percentage of missing values by each device is consistent with their representation in the data overall. This suggests that the missing data is likely randomly distributed and does not appear to be driven by device type.

##**5. Churned Users**

Recall that this dataset contains 82% retained users and 18% churned users. Next, we'll compare the medians of each variable for churned and retained users to see if there is any noticeable difference between the two groups. The reason for calculating the median and not the mean is to prevent outliers to affect the portrayal of a typical user. Recall that in the section **Summary Information** we've identified the existence of outliers.

In [20]:
# Calculate median values of all columns for churned and retained users
df.groupby('label').median(numeric_only=True)

ID  sessions  drives  total_sessions  n_days_after_onboarding  \
label                                                                         
churned   7477.5      59.0    50.0      164.339042                   1321.0   
retained  7509.0      56.0    47.0      157.586756                   1843.0   

          total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
label                                                                        
churned                     84.5                    11.0       3652.655666   
retained                    68.0                     9.0       3464.684614   

          duration_minutes_drives  activity_days  driving_days  
label                                                           
churned               1607.183785            8.0           6.0  
retained              1458.046141           17.0          14.0

**Insight:**

1. On average, **churned** users completed **~3 more drives** in the last month than retained users, while **retained** users were **active on more than twice as many days** over the same period.

2. The median **churned** user traveled **~200 additional kilometers** and spent about **2.5 more hours** driving in the last month compared with the median retained user.

3. Overall, **churned users appear to have concentrated their driving activity into fewer days, with trips that were longer in both distance and duration**. This may indicate a distinct usage pattern.



###Deeper Exploration

We further examine the observation that churned users tend to concentrate their driving activity into fewer days, with trips that are longer in both distance and duration. To do so, we compare churned and retained users across three metrics: median kilometers per drive, median kilometers per driving day, and median drives per driving day.

In [22]:
# Add a column to df called `km_per_drive`
df['km_per_drive'] = df['driven_km_drives'] / df['drives']

# Group by `label`, calculate the median, and isolate for km per drive
median_km_per_drive = df.groupby('label').median(numeric_only=True)[['km_per_drive']]
median_km_per_drive

km_per_drive
label                 
churned      74.109416
retained     75.014702

In [23]:
# Add a column to df called `km_per_driving_day`
df['km_per_driving_day'] = df['driven_km_drives'] / df['driving_days']

# Group by `label`, calculate the median, and isolate for km per driving day
median_km_per_driving_day = df.groupby('label').median(numeric_only=True)[['km_per_driving_day']]
median_km_per_driving_day

km_per_driving_day
label                       
churned           697.541999
retained          289.549333

In [24]:
# Add a column to df called `drives_per_driving_day`
df['drives_per_driving_day'] = df['drives'] / df['driving_days']

# Group by `label`, calculate the median, and isolate for drives per driving day
median_drives_per_driving_day = df.groupby('label').median(numeric_only=True)[['drives_per_driving_day']]
median_drives_per_driving_day

drives_per_driving_day
label                           
churned                  10.0000
retained                  4.0625

**Insight:**
1. The median user who **churned drove 698 kilometers each day** they drove last month, which is **~240% the per-drive-day distance of retained users**. The median churned user had a similarly disproporionate number of drives per drive day compared to retained users.

2. It is clear from these figures that, regardless of whether a user churned or not, the users represented in this data are serious drivers! It would probably be safe to assume that this data does not represent typical drivers at large. Perhaps the data&mdash;and in particular the sample of churned users&mdash;contains a high proportion of long-haul truckers.

3. In consideration of how much these users drive, it would be worthwhile to recommend to Waze that they gather more data on these super-drivers. It's possible that the reason for their driving so much is also the reason why the Waze app does not meet their specific set of needs, which may differ from the needs of a more typical driver, such as a commuter.

###Relations with Device Type

Finally, we'll examine whether there is an imbalance in how many users churned by device type. We'll calculate the percentage of each device type for each group, churned and retained.

In [26]:
# For each label, calculate the percentage of Android users and iPhone users
df.groupby('label')['device'].value_counts(normalize=True)

label     device 
churned   iPhone     0.648659
          Android    0.351341
retained  iPhone     0.644393
          Android    0.355607
Name: proportion, dtype: float64

**Insight:**
* The proportion of iPhone and Android users is consistent across both the churned and retained groups and closely matches the distribution in the overall dataset. This indicates no apparent association between device type and the likelihood of user churn.

### **6. Key Insights**

1. This dataset includes 14,999 users, of whom 82% are retained and 18% have churned.

2. The dataset contains 12 unique variables, with data types including objects, floats, and integers.

3. The label column has 700 missing values, with no evidence to suggest the missingness is non-random. These missing values account for less than 5% of the total rows.

4. On average, churned users completed ~3 more drives in the last month than retained users.

5. Retained users used the app on more than twice as many days as churned users during the last month.

6. The median churned user drove ~200 additional kilometers and spent about 2.5 more hours driving in the last month compared with the median retained user.

7. Churned users concentrated more drives into fewer days, and their trips were longer in both distance and duration. This pattern may indicate a user profile worth further investigation.

8. The median churned user drove 698 kilometers per driving day in the last month, which is roughly 240% of the per–driving-day distance of retained users.

9. Regardless of churn status, users in this dataset drive extensively, suggesting that the data may not be representative of typical drivers in the general population.
